File: SnakeBackend/main/utils/Snake.py
import math

from .Point import Point


class Snake:
    def __init__(self, start_x, start_y, start_direction):
        self.angleToItem = 0
        self.distanceToItem = 0
        self.blocks = []
        self.direction = start_direction
        self.current_direction = start_direction
        self.current_direction_top_blocked = -1
        self.current_direction_right_blocked = -1
        self.current_direction_bottom_blocked = -1
        self.current_direction_left_blocked = -1
        self.blocks.append(Point(start_x, start_y))
        self.blocks.append(self.switch_init_second_block(start_x, start_y, start_direction))

    def switch_init_second_block(self, x, y, direction):
        # top, right, bottom, left
        switcher = {0: Point(x, y + 1), 1: Point(x - 1, y), 2: Point(x, y - 1), 3: Point(x + 1, y)}
        return switcher.get(direction, Point(x, y))

    def update_item_props(self, item):
        self.angleToItem = math.degrees(math.atan2(item.y - self.blocks[0].y, item.x - self.blocks[0].x))
        self.distanceToItem = math.sqrt((item.x - self.blocks[0].x) ** 2 + (item.y - self.blocks[0].y) ** 2)

    def update_head(self, direction, head):
        # top, right, bottom, left
        switcher = {0: Point(head.x, head.y - 1), 1: Point(head.x + 1, head.y), 2: Point(head.x, head.y + 1),
                    3: Point(head.x - 1, head.y)}
        return switcher.get(direction, Point(head.x, head.y))

    # updates the snake and return false if the snake is crushed and true if everything is ok
    def update_snake(self, direction, item):
        # check if the direction we are heading to is blocked.
        # if it is blocked then return false
        if self.current_direction_top_blocked == 1 or self.current_direction_right_blocked == 1 or \
          self.current_direction_bottom_blocked == 1 or self.current_direction_left_blocked == 1:
            return False
        # check if the selected direction is on the opposite of the current direction. If so the set the selected
        # direction to the current direction
        if abs(self.current_direction - direction) == 2:
            direction = self.current_direction
        if direction == 0 and self.current_direction_top_blocked == 1:
            return False
        if direction == 1 and self.current_direction_right_blocked == 1:
            return False
        if direction == 2 and self.current_direction_bottom_blocked == 1:
            return False
        if direction == 3 and self.current_direction_left_blocked == 1:
            return False
        self.update_item_props(item)
        for i in range(len(self.blocks)):
            if i == 0:
                continue
            self.blocks[i] = self.blocks[i - 1]

        self.blocks[0] = self.update_head(direction)
        self.current_direction = direction
        return True

    def repr_json(self):
        return dict(angleToItem=self.angleToItem, distanceToItem=self.distanceToItem, blocks=self.blocks,
                    direction=self.direction, current_direction=self.current_direction,
                    current_direction_top_blocked=self.current_direction_top_blocked,
                    current_direction_right_blocked=self.current_direction_right_blocked,
                    current_direction_bottom_blocked=self.current_direction_bottom_blocked,
                    current_direction_left_blocked=self.current_direction_left_blocked)

File: api/src/models/Snake.ts
import {Point} from "./Point";
import {Item} from "./Item";

/**
 * Snake class
 */
export class Snake {
  angleToItem: number;
  distanceToItem: number;
  blocks: Point[];
  current_direction: number;
  current_direction_top_blocked: number;
  current_direction_right_blocked: number;
  current_direction_bottom_blocked: number;
  current_direction_left_blocked: number;
  id: number;

  /**
   * constructor for initialising the snake
   * @param start_x
   * @param start_y
   * @param start_direction
   * @param id
   */
  constructor(start_x: number, start_y: number, start_direction: number, id: number) {
    this.angleToItem = 0;
    this.distanceToItem = 0;
    this.blocks = [];
    this.current_direction = start_direction;
    this.current_direction_top_blocked = -1;
    this.current_direction_right_blocked = -1;
    this.current_direction_bottom_blocked = -1;
    this.current_direction_left_blocked = -1;
    this.id = id;
    this.blocks.push(new Point(start_x, start_y));
    this.blocks.push(this.switch_init_second_block(start_x, start_y, start_direction));
  }

  /**
   * helper method to set the second blocked of a snake based on the current direction of the snake.
   * @param x
   * @param y
   * @param direction
   */
  switch_init_second_block(x, y, direction) {
    // top, right, bottom, left
    switch (direction) {
      case 0:
        return new Point(x, y + 1);
      case 1:
        return new Point(x - 1, y);
      case 2:
        return new Point(x, y - 1);
      case 3:
        return new Point(x + 1, y);
    }
  }

  /**
   * for the ai it is important to get the snakes angle between the snake and the item and the x axis. Also the
   * distance to the item is a useful feature for predictions. This will not influence the game.
   * @param item
   */
  update_item_props(item: Item) {
    this.angleToItem = (Math.atan2(item.position.y - this.blocks[0].y, item.position.x - this.blocks[0].x)) * (180 / Math.PI);
    this.distanceToItem = Math.sqrt((item.position.x - this.blocks[0].x) ** 2 + (item.position.y - this.blocks[0].y) ** 2);
  }

  /**
   * this method updates the current position of the head.
   * @param direction
   * @param head
   */
  update_head(this, direction, head) {
    // top, right, bottom, left
    switch (direction) {
      case 0:
        return new Point(head.x, head.y - 1);
      case 1:
        return new Point(head.x + 1, head.y);
      case 2:
        return new Point(head.x, head.y + 1);
      case 3:
        return new Point(head.x - 1, head.y);
    }
  }

  /**
   * updates the snake and return false if the snake is crushed and true if everything is ok
   * @param direction
   * @param item
   */
  update_snake(direction: number, item: Item) {
    // check if the selected direction is on the opposite of the current direction. If so the set the selected
    // direction to the current direction
    if (Math.abs(this.current_direction - direction) === 2) {
      direction = this.current_direction;
    }
    if (direction === 0 && this.current_direction_top_blocked === 1) {
      return true
    }
    if (direction === 1 && this.current_direction_right_blocked === 1) {
      return true;
    }
    if (direction === 2 && this.current_direction_bottom_blocked === 1) {
      return true;
    }
    if (direction === 3 && this.current_direction_left_blocked == 1) {
      return true;
    }

    // eat the item if coordinates of head and item are the same
    if (item.position.x === this.blocks[0].x && item.position.y === this.blocks[0].y) {
      this.blocks.push(this.blocks[this.blocks.length - 1]);
    }

    this.update_item_props(item);
    let blocksBeforeUpdated = this.blocks.slice(0);
    for (let i = 0; i < this.blocks.length; i++) {
      if (i === 0) continue;
      // check if the item is eaten and then don´t move the block
      if (item.position.x === this.blocks[0].x && item.position.y === this.blocks[0].y && i === this.blocks.length - 1) continue;
      this.blocks[i] = blocksBeforeUpdated[i - 1];
    }

    // update head
    this.blocks[0] = this.update_head(direction, this.blocks[0]);
    this.current_direction = direction;

    return false;
  }
}
